### PR TITLE
multizone support with siwave

### PR DESCRIPTION
**Pull request by svandenb-dev**
New feature multizone support with Siwave

### CODE REVIEW

File: pyaedt/circuit.py [hunk 1]
```diff
@@ -1577,6 +1573,7 @@ def connect_circuit_models_from_multi_zone_cutout(
             Distance increment for adding models.
 
         Examples
+        --------
         command to get the input arguments.
```
Comment (PipKat) [inline pyaedt/circuit.py:1577]: This should be a complete sentence that introduces the code example.

File: pyaedt/circuit.py [hunk 2]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
         shutil.copy(latest_file, filepath)
         filename = os.path.basename(latest_file)
         return os.path.join(filepath, filename)
+
+    @pyaedt_function_handler()
+    def connect_circuit_models_from_multi_zone_cutout(
+        self, project_connections, edb_zones_dict, ports=None, schematic_units="mm", model_inc=50
+    ):
+        """Connect circuit model from multi zone clipped project.
```
Comment (PipKat) [inline pyaedt/circuit.py:1560]:
```suggestion
        """Connect circuit model from multi-zone clipped project.
```

File: pyaedt/circuit.py [hunk 3]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
...
+    @pyaedt_function_handler()
+    def connect_circuit_models_from_multi_zone_cutout(
+        self, project_connections, edb_zones_dict, ports=None, schematic_units="mm", model_inc=50
+    ):
+        """Connect circuit model from multi-zone clipped project.
+
+        Parameters
+        ----------
+        project_connections : dic[str][str]
+            project connections dictionarry return from command edb.get_connected_ports_from_multizone_cutout():
```
Comment (PipKat) [inline pyaedt/circuit.py:1565]:
```suggestion
            Dictionary of project connections returned from the
            ``edb.get_connected_ports_from_multizone_cutout()`` method.       
```
We should probably use the :func: direction to make the method a link. This is true for the other descriptions as well.

File: pyaedt/circuit.py [hunk 4]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
...
+        edb_zones_dict : dict[str][EDB PolygonData]
+            dictionary returned by command edb.copy_zones()
```
Comment (PipKat) [inline pyaedt/circuit.py:1567]:
```suggestion
            Dictionary of zones returned by the ``edb.copy_zones()`` method.
```

File: pyaedt/circuit.py [hunk 5]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
...
+        ports : dict[str][str]
+            dictionary return from command edb.cutout_multizone_layout()
```
Comment (PipKat) [inline pyaedt/circuit.py]:
```suggestion
            Dictionary of multizones returned from the ``edb.cutout_multizone_layout()``
            method.
```
Is this a dictionary of ports or multizones?

File: pyaedt/circuit.py [hunk 6]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
...
+        schematic_units : str
+            units used for shcmematic, "mm", in". Default is "mm",
```
Comment (PipKat) [inline pyaedt/circuit.py:1573]:
```suggestion
            Units for the schematic, such as ``"mm"`` or ``"in"``. The
            default is ``"mm"``.
```

File: pyaedt/circuit.py [hunk 7]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
...
+
+        Parameters
+        ----------
+        project_connections : dic[str][str]
+            project connections dictionarry return from command edb.get_connected_ports_from_multizone_cutout():
+        edb_zones_dict : dict[str][EDB PolygonData]
+            dictionary returned by command edb.copy_zones()
+        ports : dict[str][str]
+            dictionary return from command edb.cutout_multizone_layout()
+        schematic_units : str
```
Comment (PipKat) [inline pyaedt/circuit.py:1572]:
```suggestion
        schematic_units : str, optional
```

File: pyaedt/circuit.py [hunk 8]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
...
+            units used for shcmematic, "mm", in". Default is "mm",
+        model_inc : float
+            Distance increment for adding models.
```
Comment (PipKat) [inline pyaedt/circuit.py:1575]:
```suggestion
        model_inc : float, optional
            Distance increment for adding models. The default is ``50``.
```

File: pyaedt/circuit.py [hunk 9]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
...
+        command to get the input arguments.
+        >>> edb = Edb(edb_file)
+        >>> edb_zones = edb.copy_zones(r"C:\Temp\test")
+        >>> defined_ports, terminals_info = edb.cutout_multizone_layout(edb_zones, common_reference_net)
+        >>> project_connexions = edb.get_connected_ports_from_multizone_cutout(terminals_info)
+
+        Returns
+        -------
+        bool
+            ``True`` when succeeded, ``False`` if failed.
```
Comment (PipKat) [inline pyaedt/circuit.py:1591]:
```suggestion
            ``True`` when succeessful, ``False`` when failed.
```

File: pyaedt/circuit.py [hunk 10]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
...
+                            model_pin = next(pin for pin in model.pins if pin.name == port_name)
+                            if model_pin:
+                                self.modeler.components.create_interface_port(port_name, model_pin.location)
+            self.save_project()
+            return True
+        return False
+
+    @pyaedt_function_handler()
+    def import_edb_in_circuit(self, edb_path):
+        """Import EDB design inside Circuit project.
```
Comment (PipKat) [inline pyaedt/circuit.py:1640]:
```suggestion
        """Import an EDB design inside a Circuit project.
```

File: pyaedt/circuit.py [hunk 11]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
...
+
+        Parameters
+        ----------
+        edb_path : str
+            path of EDB file to be copied.
```
Comment (PipKat) [inline pyaedt/circuit.py:1645]:
```suggestion
            Path of the EDB file to copy.
```

File: pyaedt/circuit.py (hunk 1 above)
Comment (svandenb-dev) [reply pyaedt/circuit.py]: @PipKat thanks I made some changes I hope it's clear. This method is actually not easy to explain.

File: pyaedt/circuit.py [hunk 12]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
...
+
+        Returns
+        -------
+            Hfss3DLayout component instance.
```
Comment (PipKat) [inline pyaedt/circuit.py:1649]: I think you left-align use a link to the object here.

File: pyaedt/edb.py [hunk 13]
```diff
@@ -3356,3 +3358,145 @@ def calculate_initial_extent(self, expansion_factor):
         max_width = max_width * expansion_factor
         self.logger.info("The W factor is {}, The initial extent = {:e}".format(expansion_factor, max_width))
         return max_width
+
+    @pyaedt_function_handler()
+    def copy_zones(self, working_directory=None):
+        """Copy multizone EDB project to one new edb per zone.
```
Comment (PipKat) [inline pyaedt/edb.py:3364]:
```suggestion
        """Copy multizone EDB projects to one new EDB per zone.
```

File: pyaedt/edb.py [hunk 14]
```diff
@@ -3356,3 +3358,145 @@ def calculate_initial_extent(self, expansion_factor):
...
+
+        Parameters
+        ----------
+        working_directory : str
+            Directory path where all EDB project are copied, if empty will use the current EDB project.
```
Comment (PipKat) [inline pyaedt/edb.py:3369]:
```suggestion
            Directory path where all EDB project are copied. The default is ``None``,
            in which case the current EDB project is used.
```

File: pyaedt/edb.py [hunk 15]
```diff
@@ -3356,3 +3358,145 @@ def calculate_initial_extent(self, expansion_factor):
...
+
+        Returns
+        -------
+           dict[str](EDB PolygonData)
+           Return a dictionary with edb path as key and EDB polygon Data defining the region.
```
Comment (PipKat) [inline pyaedt/edb.py:3374]:
```suggestion
       dict[str](EDB PolygonData)
           Dictionary with the EDB path as the key and EDB ``PolygonData`` defining the region.
```

File: pyaedt/edb.py [hunk 16]
```diff
@@ -3356,3 +3358,145 @@ def calculate_initial_extent(self, expansion_factor):
...
+                working_directory, "{}_{}".format(zone.GetId(), os.path.basename(self.edbpath))
+            )
+            shutil.copytree(self.edbpath, edb_zone_path)
+            poly_data = zone.GetPolygonData()
+            edb_zones[edb_zone_path] = poly_data
+        return edb_zones
+
+    @pyaedt_function_handler()
+    def cutout_multizone_layout(self, zone_dict, common_reference_net):
+        """Create multizone project cutout.
```
Comment (PipKat) [inline pyaedt/edb.py:3401]:
```suggestion
        """Create a multizone project cutout.
```

File: pyaedt/edb.py [hunk 17]
```diff
@@ -3356,3 +3358,145 @@ def calculate_initial_extent(self, expansion_factor):
...
+        Parameters
+        ----------
+        zone_dict : dict[str](EDB PolygonData)
+            Dictionary with EDB path as key and EDB PolygonData as value defining the zone region.
+            This dictionary is returned from the command copy_zones():
+            >>> edb = Edb(edb_file)
+            >>> zone_dict = edb.copy_zones(r"C:\Temp\test")
+
+        common_reference_net : str
+            the common reference net name. This net name must be provided to provide a valid project.
```
Comment (PipKat) [inline pyaedt/edb.py:3412]:
```suggestion
            Common reference net name. A net name must be specified to provide a valid project.
```

File: pyaedt/edb.py [hunk 18]
```diff
@@ -3356,3 +3358,145 @@ def calculate_initial_extent(self, expansion_factor):
...
+
+    @pyaedt_function_handler()
+    def cutout_multizone_layout(self, zone_dict, common_reference_net):
+        """Create multizone project cutout.
+
+        Parameters
+        ----------
+        zone_dict : dict[str](EDB PolygonData)
+            Dictionary with EDB path as key and EDB PolygonData as value defining the zone region.
+            This dictionary is returned from the command copy_zones():
```
Comment (PipKat) [inline pyaedt/edb.py:3407]:
```suggestion
            Dictionary with an EDB path as the key and EDB ``PolygonData`` as the values
            defining the zone region. This dictionary is returned from the
            ``copy_zones()`` method.
```

File: pyaedt/edb.py [hunk 19]
```diff
@@ -3356,3 +3358,145 @@ def calculate_initial_extent(self, expansion_factor):
...
+            >>> zone_dict = edb.copy_zones(r"C:\Temp\test")
+
+        common_reference_net : str
+            the common reference net name. This net name must be provided to provide a valid project.
+
+        Returns
+        -------
+        dict[str][str] , dict[str][str]
+        first dictionary defined_ports with edb name as key and existing port name list as value
+        second dictionary terminals with edb name as key and created ports name on clipped signal nets
```
Comment (PipKat) [inline pyaedt/edb.py:3418]:
```suggestion
        First dictionary defined_ports with the EDB name as the key and existing port name list as values.
        Second dictionary terminals with the EDB name as the key and created ports name on clipped signal nets.
```
I'm not sure how to suggest rewriting these descriptions, but they are rather choppy and more codelike than in plain English. For example, defined_ports as written seens like a parameter name that should be in double back ticks. However, we prefer simple descriptions that read like a sentence from a book.

File: pyaedt/edb.py [hunk 20]
```diff
@@ -3356,3 +3358,145 @@ def calculate_initial_extent(self, expansion_factor):
...
+            )
+            if edb_terminals_info:
+                terminals[os.path.splitext(os.path.basename(edb_path))[0]] = edb_terminals_info
+            edb.save_edb()
+            edb.close_edb()
+        return defined_ports, terminals
+
+    @pyaedt_function_handler()
+    def get_connected_ports_from_multizone_cutout(self, terminal_info_dict):
+        """Return connected port list from clipped multizone layout.
```
Comment (PipKat) [inline pyaedt/edb.py:3441]:
```suggestion
        """Get the list of connected ports from the clipped multizone layout.
```

File: pyaedt/edb.py [hunk 21]
```diff
@@ -3356,3 +3358,145 @@ def calculate_initial_extent(self, expansion_factor):
...
+
+        Parameters
+            terminal_info_dict : dict[str][str]
+                dictionary terminals with edb name as key and created ports name on clipped signal nets.
+                Dictionary is generated by the command cutout_multizone_layout:
```
Comment (PipKat) [inline pyaedt/edb.py:3446]:
```suggestion
                Dictionary of terminals with the EDB name as the key and the names
                of ports created on clipped signal nets. This dictionary is generated
                by the ``cutout_multizone_layout`` method.
```

File: pyaedt/edb_core/hfss.py [hunk 22]
```diff
@@ -1024,6 +1024,84 @@ def create_lumped_port_on_net(
...
+        Parameters
+        ----------
+        nets : list, optional
+            List of nets, str or Edb net.
+
+        reference_net : str, Edb net.
+             Name or Edb reference net.
+
+        user_defined_extent : [x, y], EDB PolygonData
+            Use this point list or PolygonData object to check if ports are at this polygon border.
```
Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1040]:
```suggestion
            Point list or ``PolygonData`` object to check if ports are at this polygon border.
```

File: pyaedt/edb_core/hfss.py [hunk 23]
```diff
@@ -1024,6 +1024,84 @@ def create_lumped_port_on_net(
...
+
+        Returns
+        -------
+        [[str]]
+            Nested list of str, with net name as first value, X value for point at border, Y value for point at border,
+            and terminal name.
```
Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1046]:
```suggestion
            Nested list of strings, with net name as the first value, the X value for
            the point at the border as the second value, the Y value for the point
            at the border as the third value, and the terminal name as the last value.
```

File: pyaedt/edb_core/hfss.py [hunk 24]
```diff
@@ -1024,6 +1024,84 @@ def create_lumped_port_on_net(
...
+        if not isinstance(nets, list):
+            if isinstance(nets, str):
+                nets = list(self._pedb.nets.signal.values())
+        else:
+            nets = [self._pedb.nets.signal[net] for net in nets]
+        if nets:
+            if isinstance(reference_net, str):
+                reference_net = self._pedb.nets[reference_net]
+            if not reference_net:
+                self._logger.error("No reference net provided for creating port")
```
Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1057]:
```suggestion
                self._logger.error("No reference net is provided for creating the port.")
```
Error messages should end with a period as shown. Please fix here and in any tests.

File: pyaedt/edb_core/hfss.py [hunk 25]
```diff
@@ -1024,6 +1024,84 @@ def create_lumped_port_on_net(
...
+                                mid_pt_data = self._edb.Geometry.PointData(
+                                    self._edb.Utility.Value(mid_point[0]), self._edb.Utility.Value(mid_point[1])
+                                )
+                                ref_prim = [
+                                    prim
+                                    for prim in reference_net.primitives
+                                    if prim.polygon_data.PointInPolygon(mid_pt_data)
+                                ]
+                                if not ref_prim:
+                                    self._logger.warning("No reference primitive found in port vicinity")
```
Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1097]:
```suggestion
                                    self._logger.warning("No reference primitive is found in the port vicinity.")
```
End message with a period and fix in any tests.

File: pyaedt/modeler/geometry_operators.py [hunk 26]
```diff
@@ -1551,7 +1551,7 @@ def v_angle_sign_2D(va, vb, right_handed=True):
 
     @staticmethod
     @pyaedt_function_handler()
-    def point_in_polygon(point, polygon):
+    def point_in_polygon(point, polygon, tolerance=1e-8):
         """Determine if a point is inside or outside a polygon, both located on the same plane.
```
Comment (PipKat) [inline pyaedt/modeler/geometry_operators.py:1555]: Can we better explain this part: "both located on the same plane." Both what? Are we referring to the pain and polygon?

File: pyaedt/modeler/geometry_operators.py [hunk 27]
```diff
@@ -1560,6 +1560,8 @@ def point_in_polygon(point, polygon):
             List of ``[x, y]`` coordinates.
         polygon : List
             [[x1, x2, ..., xn],[y1, y2, ..., yn]]
+        tolerance : float
+            tolerance used for the algorithm. Default value is 1e-8.
```
Comment (PipKat) [inline pyaedt/modeler/geometry_operators.py:1564]:
```suggestion
            Tolerance for the algorithm. The default value is ``1e-8``.
```

File: pyaedt/modeler/geometry_operators.py [hunk 28]
```diff
@@ -1569,7 +1571,7 @@ def point_in_polygon(point, polygon):
             - ``1`` When the point is inside the polygon.
         """
         # fmt: off
-        tol = 1e-8
+        tol = tolerance
         if len(point) != 2:  # pragma: no cover
             raise ValueError("point must be a list in the form [x, y]")
```
Comment (PipKat) [inline pyaedt/modeler/geometry_operators.py:1576]:
```suggestion
            raise ValueError("Point must be a list in the form [x, y].")
```
Start with a capital letter and conclude the complete sentence with a period.

File: pyaedt/circuit.py [hunk 29]
```diff
@@ -1551,3 +1552,107 @@ def browse_log_file(self, filepath=None):
         shutil.copy(latest_file, filepath)
         filename = os.path.basename(latest_file)
         return os.path.join(filepath, filename)
+
+    @pyaedt_function_handler()
+    def connect_circuit_models_from_multi_zone_cutout(
+        self, project_connections, edb_zones_dict, ports=None, schematic_units="mm", model_inc=50
+    ):
+        """Connect circuit model from multi-zone clipped project.
```
Comment (PipKat) [inline pyaedt/circuit.py:1560]:
```suggestion
        """Connect circuit model from a multizone clipped project.
```

File: pyaedt/edb.py [hunk 30]
```diff
@@ -3356,3 +3358,145 @@ def calculate_initial_extent(self, expansion_factor):
...
+                >>> edb = Edb(edb_file)
+                >>> edb_zones = edb.copy_zones(r"C:\Temp\test")
+                >>> defined_ports, terminals_info = edb.cutout_multizone_layout(edb_zones, common_reference_net)
+                >>> project_connexions = get_connected_ports(terminals_info)
+
+        Returns
+        -------
+
```
Comment (PipKat) [inline pyaedt/edb.py:3454]: Need to add the return value.

File: pyaedt/edb_core/hfss.py [hunk 31]
```diff
@@ -1024,6 +1024,84 @@ def create_lumped_port_on_net(
...
         return True
 
+    @pyaedt_function_handler()
+    def create_vertical_circuit_port_on_clipped_traces(self, nets=None, reference_net=None, user_defined_extent=None):
+        """Create an edge port on clipped signal traces.
+
+        Parameters
+        ----------
+        nets : list, optional
+            List of nets, str or Edb net.
```
Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1034]: I think we should clarify this. Is the data type a list or a string? If so, it should say so in the first line, with the parameter name. The description would then be something like this: "String of one net or EDB net or a list of multiple nets or EDB nets.

File: pyaedt/edb_core/hfss.py [hunk 32]
```diff
@@ -1024,6 +1024,84 @@ def create_lumped_port_on_net(
...
+
+        reference_net : str, Edb net.
+             Name or Edb reference net.
```
Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1037]:
```suggestion
             Name or EDB reference net.
```
I also think this needs some work. Edb net is not a data type (or is it an object that should be linked?)

**Review by PipKat — CHANGES_REQUESTED** [review]

**Review by PipKat — CHANGES_REQUESTED** [review]
Doc review of new or changed docstrings

File: pyaedt/circuit.py (hunk 3 above)
Comment (svandenb-dev) [reply pyaedt/circuit.py:1565]: ok thanks

File: pyaedt/circuit.py (hunk 5 above)
Comment (svandenb-dev) [reply pyaedt/circuit.py]: It is multizones

File: pyaedt/edb_core/hfss.py (hunk 31 above)
Comment (svandenb-dev) [reply pyaedt/edb_core/hfss.py]: Yes I agree. Done

**Review by maxcapodi78 — APPROVED** [review]
Great

File: pyaedt/circuit.py [hunk 33]
```diff
@@ -1551,3 +1552,114 @@ def browse_log_file(self, filepath=None):
...
+        ----------
+        project_connections : dic[str][str]
+            Dictionary of project connections returned from the
+            ``edb.get_connected_ports_from_multizone_cutout()`` method.
+        edb_zones_dict : dict[str][EDB PolygonData]
+            Dictionary of zones returned by the ``edb.copy_zones()`` method.
+        ports : dict[str][str]
+            dictionary return from command edb.cutout_multizone_layout(). These ports are the ones created before
+            processing the multizone clipping. Like for instance ports created on components resulting from previous
+            automated workflow execution.
```
Comment (PipKat) [inline pyaedt/circuit.py:1572]:
```suggestion
            Dictionary of ports return from the ``edb.cutout_multizone_layout()`` method.
            These ports are the ones created before processing the multizone clipping.
            For example, the dictionary would include instance ports created on components
            resulting from previous execution of the automated workflow.
```

File: pyaedt/circuit.py [hunk 34]
```diff
@@ -1551,3 +1552,114 @@ def browse_log_file(self, filepath=None):
...
+        Returns
+        -------
+        bool
+            ``True`` when succeessful, ``False`` when failed.
+
+        Examples
+        --------
+        These commands show how to get input arguments described in this method:
+        - project_connections
+        - edb_zone_dict
```
Comment (PipKat) [inline pyaedt/circuit.py:1588]:
```suggestion
        This code example shows how to get these input arguments for the
        method:
 
        - ``project_connections``
        - ``edb_zone_dict``
```

File: pyaedt/edb.py [hunk 35]
```diff
@@ -3356,3 +3358,147 @@ def calculate_initial_extent(self, expansion_factor):
         max_width = max_width * expansion_factor
         self.logger.info("The W factor is {}, The initial extent = {:e}".format(expansion_factor, max_width))
         return max_width
+
+    @pyaedt_function_handler()
+    def copy_zones(self, working_directory=None):
+        """Copy multizone EDB project to one new edb per zone.
```
Comment (PipKat) [inline pyaedt/edb.py:3364]:
```suggestion
        """Copy the multizone EDB project to one new edb per zone.
```
I see you didn't like "edb" capitalized in the latter half of the sentence. I'd like to learn why--in case I'm doing something wrong when making suggestions!

File: pyaedt/edb.py [hunk 36]
```diff
@@ -3356,3 +3358,147 @@ def calculate_initial_extent(self, expansion_factor):
...
+
+        Parameters
+        ----------
+        working_directory : str
+            Directory path where all EDB project are copied, if empty will use the current EDB project.
```
Comment (PipKat) [inline pyaedt/edb.py:3369]: For this one, too, you resolved the suggestion without making it. My suggestion follows what we set up as the standard format. Can you explain why you didn't accept the change.

File: pyaedt/edb.py [hunk 37]
```diff
@@ -3356,3 +3358,147 @@ def calculate_initial_extent(self, expansion_factor):
...
+
+    @pyaedt_function_handler()
+    def cutout_multizone_layout(self, zone_dict, common_reference_net):
+        """Create multizone project cutout.
+
+        Parameters
+        ----------
+        zone_dict : dict[str](EDB PolygonData)
+            Dictionary with EDB path as key and EDB PolygonData as value defining the zone region.
+            This dictionary is returned from the command copy_zones():
```
Comment (PipKat) [inline pyaedt/edb.py:3407]:
```suggestion
            Dictionary with the EDB path as the key and EDB polygon data as the value
            defining the zone region. This dictionary is returned from ``copy_zones()`` method:
```

File: pyaedt/edb.py [hunk 38]
```diff
@@ -3356,3 +3358,147 @@ def calculate_initial_extent(self, expansion_factor):
...
+            >>> edb = Edb(edb_file)
+            >>> zone_dict = edb.copy_zones(r"C:\Temp\test")
+
+        common_reference_net : str
+            the common reference net name. This net name must be provided to provide a valid project.
```
Comment (PipKat) [inline pyaedt/edb.py:3412]:
```suggestion
            Common reference net name. This net name must be specified to provide a valid project.
```
We don't want to "provide" verb forms in one sentence. It's awkward.

File: pyaedt/edb.py (hunk 20 above)
Comment (PipKat) [reply pyaedt/edb.py:3443]: Again, I'm not sure why you didn't accept my change. If the method uses "Get", it's best to use "Get in the description". In general, you want to avoid long strings of adjectives--no more than 2, which is why I suggested the change to "Get the list of ....." In this case, you have only two adjectives (connected and port), so we could say Get the connected port list...."

File: pyaedt/edb.py [hunk 39]
```diff
@@ -3356,3 +3358,147 @@ def calculate_initial_extent(self, expansion_factor):
...
+        project_connexions = self._get_connected_ports_from_multizone_cutout(terminals)
+        return defined_ports, project_connexions
+
+    @pyaedt_function_handler()
+    def _get_connected_ports_from_multizone_cutout(self, terminal_info_dict):
+        """Return connected port list from clipped multizone layout.
+
+        Parameters
+            terminal_info_dict : dict[str][str]
+                dictionary terminals with edb name as key and created ports name on clipped signal nets.
```
Comment (PipKat) [inline pyaedt/edb.py:3447]:
```suggestion
                Dictionary of terminals with the EDB name as the key and the created ports name on clipped signal nets.
```
This "created ports name" is an awkward construction. It seems like it should either be "... and the names of the created ports on clipped signal nets." or "... and the name for ports created on clipped signal nets."

File: pyaedt/edb.py [hunk 40]
```diff
@@ -3356,3 +3358,147 @@ def calculate_initial_extent(self, expansion_factor):
...
+                Dictionary is generated by the command cutout_multizone_layout:
```
Comment (PipKat) [inline pyaedt/edb.py:3448]:
```suggestion
                This dictionary is generated by the ``cutout_multizone_layout()`` method:
```

File: pyaedt/edb.py [hunk 41]
```diff
@@ -3356,3 +3358,147 @@ def calculate_initial_extent(self, expansion_factor):
...
+                >>> edb = Edb(edb_file)
+                >>> edb_zones = edb.copy_zones(r"C:\Temp\test")
+                >>> defined_ports, terminals_info = edb.cutout_multizone_layout(edb_zones, common_reference_net)
+                >>> project_connexions = get_connected_ports(terminals_info)
+
+        Returns
+        -------
+        list[str]
+            list of connected ports.
```
Comment (PipKat) [inline pyaedt/edb.py:3457]:
```suggestion
            List of connected ports.
```
The data type in the first line is lower case but the description always starts with a capital letter.

File: pyaedt/edb_core/hfss.py [hunk 42]
```diff
@@ -1024,6 +1024,84 @@ def create_lumped_port_on_net(
...
         return True
 
+    @pyaedt_function_handler()
+    def create_vertical_circuit_port_on_clipped_traces(self, nets=None, reference_net=None, user_defined_extent=None):
+        """Create an edge port on clipped signal traces.
+
+        Parameters
+        ----------
+        nets : list, optional
+            String of one net or EDB net or a list of multiple nets or EDB nets.
```
Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1034]: In line 1033, I think we need to indicate that the data type can be a string or list.

Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1034]:
```suggestion
            String of one net or EDB net, or a list of multiple nets or EDB nets.
            The default is ``None``.
```

File: pyaedt/edb_core/hfss.py [hunk 43]
```diff
@@ -1024,6 +1024,84 @@ def create_lumped_port_on_net(
...
+
+        reference_net : str, Edb net.
+             Name or EDB reference net.
```
Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1037]:
```suggestion
        reference_net : str, Edb net.
             Name or EDB reference net. The default is ``None``.
```
I'm not sure why you keep resolving my suggestions without incorporating any of them. If I'm making bad suggestions, I'd like to understand why. My doc edit is intended to improve readability, which include inserting articles like these: a, an, the. Also, messages should be sentences and include the "is" helping verb where needed.

File: pyaedt/edb_core/hfss.py [hunk 44]
```diff
@@ -1024,6 +1024,84 @@ def create_lumped_port_on_net(
...
+
+        user_defined_extent : [x, y], EDB PolygonData
+            Use this point list or PolygonData object to check if ports are at this polygon border.
```
Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1040]: You resolved my suggested edit without accepting my comments--but parameter descriptions are not supposed to start with verb.

File: pyaedt/edb_core/hfss.py [hunk 45]
```diff
@@ -1024,6 +1024,84 @@ def create_lumped_port_on_net(
...
+
+        Returns
+        -------
+        [[str]]
+            Nested list of str, with net name as first value, X value for point at border, Y value for point at border,
```
Comment (PipKat) [inline pyaedt/edb_core/hfss.py:1045]:
```suggestion
            Nested list of string, with net name as first value, X value for point at border, Y value for point at border,
```
The "str" format should only be used in the first line, which provides the data type. The description itself should use words (and articles as per my previous comment that is now resolved).

**Review by PipKat — APPROVED** [review]

**Review by PipKat — APPROVED** [review]
